### PR TITLE
Feature/custom urls

### PR DESCRIPTION
**Pull request by KamranIE**

### CODE REVIEW

File: src/Acme.UI/Infrastructure/DI/AcmeComposer.cs [hunk 1]
```diff
@@ -11,11 +14,13 @@ public void Compose(Composition composition)
         {
             RegisterComponents(composition);
             RegisterServices(composition);
+            RegisterContentFinders(composition);
         }
 
         private void RegisterComponents(Composition composition)
         {
             composition.Components().Append<MemberSaveHandler>();
+                                    //.Append<AcmeRouter>();
```
Comment (ettiennegous) [inline src/Acme.UI/Infrastructure/DI/AcmeComposer.cs]: Remove commented code

File: src/Acme.UI/Infrastructure/Routing/AcmeContentFinder.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+using Umbraco.Web;
+using Umbraco.Web.Routing;
+
+namespace Acme.UI.Infrastructure.Routing
+{
+    public class AcmeContentFinder : IContentFinder
+    {
+        private readonly Dictionary<string, RouteParameters> _pathToContentType = new Dictionary<string, RouteParameters>
+        {
+            { "(?<home>dashboard)/(?<section>athlete)/(?<athleteName>[a-zA-Z0-9-]+)", new RouteParameters(typeof(Umbraco.Web.PublishedModels.Athlete), "Athlete", "//home/contentPage/contentPage")}
```
Comment (ettiennegous) [inline src/Acme.UI/Infrastructure/Routing/AcmeContentFinder.cs]: Why did you need the regex in both the athletes controller and in this file?

File: src/Acme.UI/Infrastructure/Routing/AcmeRouter.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        public AcmeRouter(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public void Initialize()
+        {
+            _logger.Info<AcmeRouter>("-->Initializing custom acme routing4");
+
+            /*RouteTable.Routes.MapRoute("LoginRoute1", "kamran/member/render_login/{returnUrl}",
```
Comment (ettiennegous) [inline src/Acme.UI/Infrastructure/Routing/AcmeRouter.cs]: Removed commented code please

File: src/Acme.UI/Infrastructure/Routing/RouteParameters.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        public string SearchNodeByName { get; private set; }
+        public string SearchNodeByNameFormatted { get; private set; }
+
+        public string RootNodeName { get; private set; }
+
+        public RouteParameters(Type dataSourceType, string searchNodeByName, string rootNodeName)
+        {
+            DataSourceType = dataSourceType;
+            SearchNodeByName = searchNodeByName;
+            SearchNodeByNameFormatted = searchNodeByName.Replace("-", " ");
```
Comment (ettiennegous) [inline src/Acme.UI/Infrastructure/Routing/RouteParameters.cs]: There is a built in function called ToUrlSegement
https://our.umbraco.com/documentation/reference/routing/request-pipeline/outbound-pipeline

I think you are taking care of dashes but what about other special characters like ampersand etc.

File: src/Acme.UI/Controllers/Business/AthletesController.cs [hunk 5]
```diff
@@ -5,38 +5,60 @@
...
 
         public ActionResult List(int? maxAthletesToDisplay)
         {
             var logginPhysioNodeId = GetLoggedInPhysioNodeId();
-            
+            var athletesForLoggedInUser = GetAthletesForLoggedInUser(logginPhysioNodeId);
             var physioAthletesViewModel = PhysioAthletesViewModel.Create(Umbraco, logginPhysioNodeId ?? 0, maxAthletesToDisplay);
             
-            if (logginPhysioNodeId != null)
+            if (athletesForLoggedInUser != null)
```
Comment (ettiennegous) [inline src/Acme.UI/Controllers/Business/AthletesController.cs]: May I suggest blocking statements at the beginning of the method,

File: src/Acme.UI/Infrastructure/Routing/AcmeContentFinder.cs (hunk 2 above)
Comment (KamranIE) [reply src/Acme.UI/Infrastructure/Routing/AcmeContentFinder.cs]: We probably need a simpler pattern in AcmeContentFinder. Here it is just trying to find the right matching record in _pathToContentType to assign the right published content (i.e. page) with the request. 

In Athlete controller we need value in the Url for athlete name to process it and get the athlete record.

File: src/Acme.UI/Infrastructure/Routing/AcmeRouter.cs (hunk 3 above)
Comment (KamranIE) [reply src/Acme.UI/Infrastructure/Routing/AcmeRouter.cs]: We don't need this class. I will remove it.

File: src/Acme.UI/Infrastructure/Routing/RouteParameters.cs (hunk 4 above)
Comment (KamranIE) [reply src/Acme.UI/Infrastructure/Routing/RouteParameters.cs]: Here the purpose is to convert the url to a name used for the node in the data hierarchy. I think "ToUrlSegment" doing the inverse of it.

Comment (KamranIE) [reply src/Acme.UI/Infrastructure/Routing/RouteParameters.cs]: As per my investigation the generated Url for "acme\Test&^%*#-_=+$%@!star223" was "test-_-star223". umbraco trims almost all special characters (except hyphen and underscore). 

In SearchNodeByNameFormatted = searchNodeByName.Replace("-", " ");
we want to replace hyphen with space so that we could match it with athlete name. I don't think umbraco would be capable to change test-_-star223 to "Test&^%*#-_=+$%@!star223" in our case. let me know what you think.

**Review by ettiennegous — APPROVED** [review]
